Fixed Factory/Label conflict

benmagruder opened the pull request:
This pull request fixes a conflict that I encountered between the Factory class and the Label class. 

The issue occurred while reading a parameter file. The two lines of code that created the issue are given below:
```
readParamCompositeOptional(in, wall());
iteratorPtr_ = iteratorFactoryPtr_->readObject(in, *this, className, isEnd);
```
I was attempting to read an optional paramComposite, followed by reading a polymorphic object by calling `Factory::readObject()`. If the optional paramComposite was omitted from the param file, then the code broke.

Here I explain why the code broke. Most objects are read by the Label class. The expected name of the object is stored in `Label::string`, and then the text that is read from the param file is stored in `Label::input_`. If the two members of Label match, then we create the object. If they don't match, then we throw an error (if the object was required) or move on to the next object we wish to read (if the object was optional). In the latter case, we will update `Label::string` and again check if it matches `Label::input_`, because `Label::input_` contains whatever was on the next line of the param file and we can't proceed in the param file until it is matched.

The problem arises when `Label::input_` contains an unmatched string and we attempt to read a polymorphic object (via `Factory::readObject`). The Factory doesn't actually know the expected string beforehand, so we can't just update `Label::string` and expect it to match `Label::input_`. Instead, Factory was implemented so that it read its input directly from the param file and then used the string to identify the object that it should create. However, `Factory::readObject()` was reading directly from the param file even if the string that it was looking for was already stored in `Label::input_`. This meant that it was reading the next item in the param file which was not the actual string that it was looking for, so the Factory could not successfully read the object it was meant to read.

To solve this, I made a relatively simple change: make Factory a friend of Label. This way, Factory can see what is stored in `Label::input_` (even though this member is private), allowing it to find the correct string regardless of whether it is stored in `Label::input_` or is the next item in the param file. This essentially overrides the default behavior of Label in the case of a polymorphic block, which seems reasonable given that Label was designed to read objects with names that are known in advance (which is not true of a polymorphic block).

benmagruder committed:
Fixed typo in MatrixParam

benmagruder committed:
Fixed typo in MatrixParam.tpp

benmagruder committed:
Merge branch 'bugfix' of benmagruder/util into bugfix

benmagruder committed:
Made Factory a friend of Label to avoid readOptional conflict

benmagruder commented:
One other note: unit tests have been checked for both util and pscfpp, and everything works correctly after these changes have been made.

dmorse commented:
1) I don't understand the purpose of the second "in >> label" statement on line 241 of Factory<T>::readObject.
If label.isMatched is true on entry, won't you read from the input file twice?

2) The relevant functions and data members of Label are all static. I don't see why you need to create an instance of that class. 

3) You should not be reading unless paramFileIo.isIoProcessor is true.

How about something closer to the following

if (paramFileIo_.isIoProcessor()) {
    if (Label::isClear()) {
         in >> commentString;
    } else {
         // You could assert that isMatched is true here, as a sanity check
         commentString = Label::input_;
         Label::input_.clear();
         Label::isClear_ = true;
         Label::isMatched = true;
    }
}

benmagruder commented:
These are all good questions – the way I implemented the fix is definitely a bit confusing in that regard. I'll explain why I did it this way, but switching to the code you wrote above instead would accomplish the same thing so we can do it whichever way you prefer.

The reason that I implemented it this way is so that we can use the infrastructure that is already built into the Label `>>` operator wherever possible. With the way that I wrote the code, we never read directly from the param file into a string, we always read into a Label instead.

The key trick embedded in the `>>` operator, and thus the reason this works, is that it does not always read new data from the param file when called; it only reads new data if `label.isClear = true`. But regardless of whether it reads new data or not, the `>>` operator updates the Label static members to the correct state based on what is stored in `input_` and `string_`, so I figured we could use that feature here as well, to be consistent with the way that other functions use this `>>` operator. 

To be specific, consider the second time that we call `in >> label`. Because of the code that precedes this call, we can guarantee at this point that `Label::isClear_ = false`, and that `label.input_ == label.string_`.  So when we call `in >> label` the second time, it results in the following code being run (part of the Label `>>` operator):
```
Label::input_.clear();
Label::isClear_ = true;
Label::isMatched_ = true;
```
(note that these three lines of code are identical to what you wrote in your comment)

Nothing else in the `>>` operator is called. So, we know for sure this call of `in >> label` does not actually read anything, and it thus does not require that paramFileIo.isIoProcessor is true.

If we implement the fix this way, we must create an instance of the Label class so that we can call `in >> label` and use the code that was written in the Label `>>` operator. 

Using the Label `>>` operator this way definitely makes the code a bit more confusing to read, because a reader would naturally assume that the `>>` operator will always read new data from the param file. If we want to maximize readability, we should implement the fix in the way that you described in your comment. But using the Label `>>` operator like I did in this Pull Request is more consistent with other code, and uses Label in the way it was designed to be used, so there are benefits to this approach as well.

dmorse commented:
Ben,
     As someone who just tried to read the code, and got confused, I think
I prefer the version that we both think is easier to understand.  Would you
be willing to implement and re-test this version?

On Wed, Jul 13, 2022 at 11:28 AM Ben Magruder ***@***.***>
wrote:

> These are all good questions – the way I implemented the fix is definitely
> a bit confusing in that regard. I'll explain why I did it this way, but
> switching to the code you wrote above instead would accomplish the same
> thing so we can do it whichever way you prefer.
>
> The reason that I implemented it this way is so that we can use the
> infrastructure that is already built into the Label >> operator wherever
> possible. With the way that I wrote the code, we never read directly from
> the param file into a string, we always read into a Label instead.
>
> The key trick embedded in the >> operator, and thus the reason this
> works, is that it does not always read new data from the param file when
> called; it only reads new data if label.isClear = true. But regardless of
> whether it reads new data or not, the >> operator updates the Label
> static members to the correct state based on what is stored in input_ and
> string_, so I figured we could use that feature here as well, to be
> consistent with the way that other functions use this >> operator.
>
> To be specific, consider the second time that we call in >> label.
> Because of the code that precedes this call, we can guarantee at this point
> that Label::isClear_ = false, and that label.input_ == label.string_. So
> when we call in >> label the second time, it results in the following
> code being run (part of the Label >> operator):
>
> Label::input_.clear();
>
> Label::isClear_ = true;
>
> Label::isMatched_ = true;
>
>
> (note that these three lines of code are identical to what you wrote in
> your comment)
>
> Nothing else in the >> operator is called. So, we know for sure this call
> of in >> label does not actually read anything, and it thus does not
> require that paramFileIo.isIoProcessor is true.
>
> If we implement the fix this way, we must create an instance of the Label
> class so that we can call in >> label and use the code that was written
> in the Label >> operator.
>
> Using the Label >> operator this way definitely makes the code a bit more
> confusing to read, because a reader would naturally assume that the >>
> operator will always read new data from the param file. If we want to
> maximize readability, we should implement the fix in the way that you
> described in your comment. But using the Label >> operator like I did in
> this Pull Request is more consistent with other code, and uses Label in the
> way it was designed to be used, so there are benefits to this approach as
> well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dmorse/util/pull/1#issuecomment-1183433211>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOB4IFL2K75RPSPFOGXDR3VT3VEFANCNFSM53MAS7IQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

benmagruder commented:
Yep, I can do that!
